### PR TITLE
#2147 Update form builder to work with v3 template object fields

### LINKED ISSUE #2147: Update form builder to work with v3 template object fields
We need to refactor form builder so that it can handle config values using string-template objects in the v3 runtime, for example:
```
{
  __type__: "var"
  __value__: "@data.name"
}
```

### PR DESCRIPTION
**Pull request by BALEHOK**
Closes #2147 

Unwrapping expressions before using them in preview.

### CODE REVIEW

File: src/components/formBuilder/FormPreview.tsx [hunk 1]
```diff
@@ -47,26 +48,12 @@ const FormPreview: React.FC<FormPreviewProps> = ({
...
-      setActiveField(firstInOrder);
-      return;
-    }
-
-    const firstInProperties = Object.keys(schema?.properties || {})[0];
-    if (firstInProperties) {
-      setActiveField(firstInProperties);
-    }
-    // eslint-disable-next-line react-hooks/exhaustive-deps -- resetting activeField only on new name
-  }, [name]);
```
Comment (BALEHOK) [inline src/components/formBuilder/FormPreview.tsx:68]: Editor inits the active field. Also `name` at ln 68 refers to a global variable, not a component property.

File: src/components/formBuilder/FormPreview.tsx [hunk 2]
```diff
@@ -47,26 +48,12 @@ const FormPreview: React.FC<FormPreviewProps> = ({
     rjsfSchema
   );
 
+  const previewSchema = useMemo(() => getPreviewValues(schema), [schema]);
```
Comment (BALEHOK) [inline src/components/formBuilder/FormPreview.tsx:51]: This is the trick.

File: src/runtime/mapArgs.ts [hunk 3]
```diff
@@ -63,7 +63,7 @@ export function isPipelineExpression(
   return isExpression(value) && value.__type__ === "pipeline";
 }
 
-export type DeferExpression = Expression<UnknownObject, "defer">;
+export type DeferExpression<TValue = string> = Expression<TValue, "defer">;
```
Comment (BALEHOK) [inline src/runtime/mapArgs.ts:66]: Default to `string` - same as `Expression`

Comment (twschiller) [reply src/runtime/mapArgs.ts:66]: IIRC, Defer doesn't support primitive values for `__value__`. (The YAML tag serializer/deserializer requires an object under the defer. Pipeline requires an array) 

What was the reason for changing this to string?

File: src/components/fields/fieldUtils.ts [hunk 4]
```diff
@@ -48,3 +51,23 @@ export function createTypePredicate(predicate: TypePredicate): TypePredicate {
     return false;
   };
 }
+
+export function getPreviewValues<TObj = UnknownObject>(obj: TObj): TObj {
+  return produce(obj, (draft) => {
```
Comment (twschiller) [inline src/components/fields/fieldUtils.ts:56]: This method is confusing to read, that it's a function within a function, and the function definition is after it's call. 

Can we move unwrapTemplateExpressions outside of getPreviewValues?

File: src/components/fields/fieldUtils.ts [hunk 5]
```diff
@@ -48,3 +51,23 @@ export function createTypePredicate(predicate: TypePredicate): TypePredicate {
...
+    unwrapTemplateExpressions(draft);
+  });
+
+  function unwrapTemplateExpressions(mutableObj: Draft<any>) {
+    if (typeof mutableObj !== "object") {
```
Comment (twschiller) [inline src/components/fields/fieldUtils.ts:61]: Arrays are also typeof object. Is that OK? Should this be using isPlainObject?

**Review by twschiller — CHANGES_REQUESTED** [review]
@BALEHOK 

The PR looks good. The only 2 things that stuck out to me were:
- The default type of Defer Expression
- That helper method definition (do we need to differentiate arrays vs. objects)?

Can chat these over on slack quickly if you prefer

File: src/components/fields/fieldUtils.ts (hunk 5 above)
Comment (BALEHOK) [reply src/components/fields/fieldUtils.ts:61]: yes, arrays are ok. here I want to quit earlier in case of a "value" type: string, number, boolean

Comment (BALEHOK) [reply src/components/fields/fieldUtils.ts:61]: In fact`isPlainObject` fails on immer `Draft` here, `Draft` is not a plain object, so it always returns `false`.

File: src/runtime/mapArgs.ts (hunk 3 above)
Comment (BALEHOK) [reply src/runtime/mapArgs.ts]: Just made the default type to the same as the default type of the `Expression`. Reverted to `UnknownObject`.

**Review by twschiller — APPROVED** [review]

**Pull request merged**